llvm/lib/IR/Instruction.cpp

```diff
@@ -920,6 +920,8 @@ bool Instruction::hasSameSpecialState(const Instruction *I2,
   if (const GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(I1))
     return GEP->getSourceElementType() ==
            cast<GetElementPtrInst>(I2)->getSourceElementType();
+  if (const CastInst *Cast = dyn_cast<CastInst>(I1))
```

Comment:
It would be good to have a PR for these changes in isolation with some non-vectoriser tests that show the improvement.

Reply:
FWIW, I couldn't find any dedicated tests for `Instruction::isSameOperationAs`, so perhaps the changes to existing tests would be sufficient?